elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex

```diff
@@ -0,0 +1,201 @@
+      |> Ecto.Multi.append(Actors.sync_provider_groups_multi(provider, actor_groups_attrs))
+      |> Actors.sync_provider_memberships_multi(provider, tuples)
+      |> Ecto.Multi.update(:save_last_updated_at, fn _effects_so_far ->
+        Auth.Provider.Changeset.sync_finished(provider)
+      end)
+      |> Domain.Repo.transaction()
+      |> case do
+        {:ok, effects} ->
+          log_effects(provider, effects)
+
+        {:error, reason} ->
+          Logger.error("Failed to sync provider",
+            provider_id: provider.id,
+            account_id: provider.account_id,
+            reason: inspect(reason)
+          )
+      end
+    else
+      {:error, {status, %{"error" => %{"message" => message}}}} ->
+        provider =
+          Auth.Provider.Changeset.sync_failed(provider, message)
+          |> Domain.Repo.update!()
+
+        log_sync_error(provider, "Microsoft Entra API returned #{status}: #{message}")
+
+      {:error, :retry_later} ->
+        message = "Microsoft Entra API is temporarily unavailable"
+
+        provider =
+          Auth.Provider.Changeset.sync_failed(provider, message)
+          |> Domain.Repo.update!()
+
+        log_sync_error(provider, message)
+
+      {:error, reason} ->
+        Logger.error("Failed syncing provider",
+          account_id: provider.account_id,
+          provider_id: provider.id,
+          reason: inspect(reason)
+        )
+    end
+  end
+
+  defp log_sync_error(provider, message) do
+    metadata = [
+      account_id: provider.account_id,
+      provider_id: provider.id,
+      reason: message
+    ]
+
+    if provider.last_syncs_failed >= 3 do
+      Logger.warning("Failed syncing provider", metadata)
+    else
+      Logger.info("Failed syncing provider", metadata)
+    end
+  end
+
+  defp list_membership_tuples(access_token, groups) do
+    Enum.reduce_while(groups, {:ok, []}, fn group, {:ok, tuples} ->
+      case MicrosoftEntra.APIClient.list_group_members(access_token, group["id"]) do
+        {:ok, members} ->
+          tuples = Enum.map(members, &{"G:" <> group["id"], &1["id"]}) ++ tuples
+          {:cont, {:ok, tuples}}
+
+        {:error, reason} ->
+          {:halt, {:error, reason}}
+      end
+    end)
+  end
+
+  # Map identity attributes from Microsoft Entra to Domain
+  defp map_identity_attrs(users) do
+    Enum.map(users, fn user ->
+      %{
+        "provider_identifier" => user["id"],
+        "provider_state" => %{
+          "userinfo" => %{
+            "email" => user["mail"]
+          }
+        },
+        "actor" => %{
+          "type" => :account_user,
+          "name" => user["displayName"]
+        }
+      }
+    end)
+  end
+
+  # Map group attributes from Microsoft Entra to Domain
+  defp map_group_attrs(groups) do
+    Enum.map(groups, fn group ->
+      %{
+        "name" => "Group:" <> group["displayName"],
+        "provider_identifier" => "G:" <> group["id"]
```

Comment:
Is the `G` for Google?

```suggestion
        "provider_identifier" => "M:" <> group["id"]
```

Reply:
I was thinking the `G` stood for Group, but let me double check that.

Reply:
`G` is for `Group`. Because google has "OU:". But if Entra doesn't have any other entities we can remove the prefix.

Reply:
I decided to keep the `G:` prefix for now.  There are organizational units in Entra, but I'm not quite clear on whether a single tenant would actually use them.  If it turns out they do then having the `G:` prefix will be good and if they don't, then the `G:` prefix won't get in the way.